entgql/extension.go

```diff
@@ -231,6 +285,299 @@ func (e *Extension) mapScalar(f *gen.Field, op gen.Op) string {
+						Name: newASTName("field"),
+						Type: newASTNamed(orderByField),
+					}),
+				},
+			},
+		)
+	}
+	return hasOrderByType
+}
+
+// addOrderByDefinitions creates a GraphQL query field for the given node, adding
+// the relevanant arguments based on whether the relay spec should be followed,
+// ordering is enabled, and/or the where filters are enabled
+func (e *Extension) addQueryDefinition(queries map[string]*ast.FieldDefinition, node *gen.Type, relaySpec bool, orderBy bool) {
+	queryName := node.Table()
+	if relaySpec && e.relayQuery != nil {
+		queryName = e.naming(e.relayQuery(queryName))
+	}
+	query := ast.NewFieldDefinition(&ast.FieldDefinition{
+		Name: newASTName(queryName),
+		Type: ast.NewList(&ast.List{
+			Type: newASTNamed(node.Name),
+		}),
+		Arguments: []*ast.InputValueDefinition{
+			ast.NewInputValueDefinition(&ast.InputValueDefinition{
+				Name: newASTName("first"),
+				Type: newASTNamed(graphql.Int.String()),
+			}),
+			ast.NewInputValueDefinition(&ast.InputValueDefinition{
+				Name: newASTName("last"),
+				Type: newASTNamed(graphql.Int.String()),
+			}),
+		},
+	})
+
+	if relaySpec {
+		query.Type = newASTNamed(node.Name + "Connection")
+		query.Arguments = append(query.Arguments,
+			ast.NewInputValueDefinition(&ast.InputValueDefinition{
+				Name: newASTName("before"),
+				Type: newASTNamed("Cursor"),
+			}),
+			ast.NewInputValueDefinition(&ast.InputValueDefinition{
+				Name: newASTName("after"),
+				Type: newASTNamed("Cursor"),
+			}))
+	}
+
+	if orderBy {
+		query.Arguments = append(query.Arguments, ast.NewInputValueDefinition(
+			&ast.InputValueDefinition{
+				Name: newASTName(e.naming("order_by")),
+				Type: newASTNamed(orderType(node.Name)),
+			},
+		))
+	}
+
+	if e.whereFilters {
+		query.Arguments = append(query.Arguments, ast.NewInputValueDefinition(
+			&ast.InputValueDefinition{
+				Name: newASTName("where"),
+				Type: newASTNamed(node.Name + "WhereInput"),
+			},
+		))
+	}
+
+	queries[queryName] = query
+}
+
+// addRelayObjectDefinitions creates the types (Connection and Edge) for the given node
+// to follow the relay spec
+func (e *Extension) addRelayObjectDefinitions(objects map[string]*ast.ObjectDefinition, node *gen.Type) {
+	conn := ast.NewObjectDefinition(&ast.ObjectDefinition{
+		Name: newASTName(node.Name + "Connection"),
+		Description: ast.NewStringValue(&ast.StringValue{
+			Value: fmt.Sprintf("%s supports the relay connection specification for node %s in the ent schema.\nGenerated by ent.", node.Name+"Connection", node.Name),
+		}),
+	})
+	conn.Fields = []*ast.FieldDefinition{
+		ast.NewFieldDefinition(&ast.FieldDefinition{
+			Name: newASTName(e.naming("total_count")),
+			Type: newASTNamed(graphql.Int.String() + "!"),
+		}),
+		ast.NewFieldDefinition(&ast.FieldDefinition{
+			Name: newASTName(e.naming("page_info")),
+			Type: newASTNamed("PageInfo!"),
+		}),
+		ast.NewFieldDefinition(&ast.FieldDefinition{
+			Name: newASTName("edges"),
```

Comment:
I prefer to have the option to make these more strict.
E.g.
make the `edges` and `node` not null. What do you think?

Reply:
Makes sense and can be done!